Translate Higher-Order Components (#77)
reactjs/pt-BR.reactjs.org: content/docs/higher-order-components.md

---
id: higher-order-components
title: Componentes de Ordem Superior
permalink: docs/higher-order-components.html
---

> Um componente de ordem superior (HOC, do inglês Higher-Order Component) é uma técnica avançada no React para reutilizar a lógica de um component. HOCs não são parte da API do React por si só. Eles pertencem a um padrão que surgiu da própria natureza de composição do React.

Author: I still used `HOC` as the abbreviation, though I'm not sure if I should have used `COS` instead.

Reviewer: Dessa maneira, aparecendo no começo do texto com o inglês do lado é perfeitamente possível. 

Por mim pode manter assim... eu acho bem desagradável quando traduzem siglas já conhecidas, vira outra coisa... é que nem ler um texto sobre DNA que os caras tratam como ADN, dá um bug mental!

}
```

`CommentList` e `BlogPost` não são identicos - eles chamam métodos diferentes de `DataSource` e renderizam saídas diferentes. Mas muito de suas implementações é igual:

- Em mount, adicione um change listener para `DataSource`.

> - Dentro do listener, chame `setState` sempre que a fonte de dados mude.

Reviewer 2: ```suggestion
- Dentro do listener, chame `setState` sempre que a fonte de dados mudar.
```

);
```

> O primeiro paramêtro é um componente envolvido. O segundo parâmetro acessa os dados que estamos interessados, dado um `DataSource` e as props atuais.

Reviewer 3: ```suggestion
O primeiro parâmetro é um componente envolvido. O segundo parâmetro acessa os dados que estamos interessados, dado um `DataSource` e as props atuais.
```

> Quando `CommentListWithSubscription` e `BlogPostWithSubscription` são renderizados, `CommentList` e `BlogPost` passarão uma prop `data` com o dado mais recente retirado de `DataSource`:

Reviewer: ```suggestion
Quando `CommentListWithSubscription` e `BlogPostWithSubscription` são renderizados, `CommentList` e `BlogPost` receberão uma prop `data` com os dados mais recentes obtidos de `DataSource`:
```

    }

    componentDidMount() {

>       // ... o qual lida com a subscrição...

Reviewer: ```suggestion
      // ... que lida com a subscrição...
```

    }

    render() {

>       // ... e renderiza o componente envolvido com os dados frescos!

Reviewer 2: ```suggestion
      // ... e renderiza o componente envolvido com os dados novos!
```

      // Note que nós passamos diretamente qualquer prop adicional
      return <WrappedComponent data={this.state.data} {...this.props} />;
    }
  };
}
```

> Note que um HOC não modifica o componente de entrada, nem utiliza herança para copiar seu comportamento. Ao invés disso, um HOC *compõe* o componente original ao *envolvê-lo* em um componente container. Um HOC é uma função pura sem efeitos colaterais.

Reviewer: ```suggestion
Note que um HOC não modifica o componente de entrada, nem utiliza herança para copiar seu comportamento. Em vez disso, um HOC *compõe* o componente original ao *envolvê-lo* (wrapping) em um componente container. Um HOC é uma função pura sem efeitos colaterais.
```

> E é isso! O componente envolvido recebe todas as props do container, junto de uma nova prop, `data`, o qual a utiliza para renderizar sua saída. O HOC não se preocupa com o como ou por que de seus dados serem usados, e o componente envolvido não se preocupa de onde os dados vieram.

Reviewer 2: ```suggestion
E é isso! O componente envolvido recebe todas as props do container, junto de uma nova prop, `data`, o qual a utiliza para renderizar sua saída. O HOC não se preocupa com o como ou o porquê de seus dados serem usados, e o componente envolvido não se preocupa de onde os dados vieram.
```

> Por `withSubscription` ser uma função normal, você pode adicionar quantos argumentos quiser. Por exemplo, você pode querer fazer o nome da prop `data` ser configurável, para continuar a isolar o HOC do componente envolvido. Ou podes aceitar um argumento que configura `shouldComponentUpdate`, ou um que configura a fonte de dados. Todos esses casos são possíveis porque o HOC tem controle total sobre como o componente é definido.

Reviewer 2: ```suggestion
Por `withSubscription` ser uma função normal, você pode adicionar quantos argumentos quiser. Por exemplo, você pode querer fazer o nome da prop `data` ser configurável, para continuar a isolar o HOC do componente envolvido. Ou você pode aceitar um argumento que configura `shouldComponentUpdate`, ou um que configura a fonte de dados. Todos esses casos são possíveis porque o HOC tem controle total sobre como o componente é definido.
```

Reviewer: ```suggestion
Por `withSubscription` ser uma função normal, você pode adicionar quantos argumentos quiser. Por exemplo, você talvez queira tornar o nome da prop `data` configurável, para isolar o HOC do componente encapsulado (wrapped component). Ou você poderia aceitar um argumento que configura `shouldComponentUpdate`, ou um que configura a fonte de dados. Todos esses casos são possíveis porque o HOC tem controle total sobre como o componente é definido.
```

> Como componentes, o contrato entre `withSubscription` e o componente envolvido é completamente baseado em props. Isso faz com que seja fácil trocar um HOC por outro, desde que eles providenciem as mesmas props para o componente envolvido. Isso pode ser útil se você mudar de bibliotecas para busca de dados, por exemplo.

Reviewer: ```suggestion
Como componentes, o contrato entre `withSubscription` e o componente encapsulado é completamente baseado em props. Isso faz com que seja fácil trocar um HOC por outro, desde que eles providenciem as mesmas props para o componente encapsulado. Isso pode ser útil se você mudar de bibliotecas para obtenção de dados, por exemplo.
```

# Não Altere o Componente Original. Use Composição. {#dont-mutate-the-original-component-use-composition}

> Resista a tentação de modificar o prototype de um componente (ou alterá-lo de qualquer outra forma) dentro de um HOC.

Reviewer: ```suggestion
Resista à tentação de modificar o prototype de um componente (ou alterá-lo de qualquer outra forma) dentro de um HOC.
```

```js
function logProps(InputComponent) {
  InputComponent.prototype.componentWillReceiveProps = function(nextProps) {
    console.log('Current props: ', this.props);
    console.log('Next props: ', nextProps);
  };
  // O fato de estarmos retornando a entrada original é uma dica de que ela sofreu mutação.
  return InputComponent;
}

// EnhancedComponent criará logs sempre que uma prop for recebida
const EnhancedComponent = logProps(InputComponent);
```

> Existem alguns problemas nisso. Primeiro, o componente de entrada não pode ser reutilizado separadamente do componente melhorado. Mais crucialmente, se você aplicar outro HOC para `EnchancedComponent` que *também* altera `componentWillReceiveProps`, a funcionalidade do primeiro HOC será sobrescrita! Esse HOC também não funcionará com componentes funcionais, os quais não possuem métodos de ciclo de vida.

Reviewer: Fiquei um pouco confuso aqui, a sobrescrição vai acontecer:

1.  Na primeira funcionalidade do HOC ?
2. Na funcionalidade do primeiro HOC ?

Author: Creio que, se você aplicar outro HOC que utilize do `componentWillReceiveProps`, isso afetará o primeiro HOC, sobrescrevendo o que acontece dentro de `componentWillReceiveProps`. Então faz mais sentido ser `na funcionalidade do primeiro HOC`.

Reviewer: Perfeito, ficou mais claro então.

> Realizar mutações em HOCs é uma abstração mal vedada - o consumidor deve saber como eles são implementados para que se possa evitar conflitos com outros HOCs.

Reviewer: "Abstração mal vedada" não dá muita informação sobre o contexto, o que acham de sermos mais diretos:

```suggestion
Realizar mutações em HOCs podem causar "vazamentos" - o consumidor deve saber como eles são implementados para evitar conflitos com outros HOCs.
```

> Ao invés de mutações, HOCs devem utilizar composição, envolvendo o componente de entrada em um componente container:

Reviewer: ```suggestion
Em vez de mutações, HOCs devem utilizar composição, encapsulando o componente de entrada em um componente container:
```

      console.log('Next props: ', nextProps);
    }
    render() {

>       // Envolve o componente de entrada em um container, sem alterá-lo. Bom!

Reviewer: ```suggestion
      // Encapsula o componente de entrada em um container, sem alterá-lo. Excelente!
```

Creio que ninguém usa "Bom!" para expressar satisfação, "excelente!" me parece mais verossímil, concordam?

Author: "Ótimo" também pode set uma boa alternativa

Reviewer: > "Ótimo" também pode ser uma boa alternativa

Com certeza!

      return <WrappedComponent {...this.props} />;
    }
  }
}
```

> Esse HOC possui a mesma funcionalidade que a sua versão com mutação e evita o potencial de ocorrer choques. Ele funciona igualmente bem com componentes funcionais e controlados. E por ser uma função pura, pode ser combinado com outros HOCs, ou até com si mesmo.

Reviewer: ```suggestion
Esse HOC possui a mesma funcionalidade que a sua versão com mutação e evita potenciais conflitos. Ele funciona igualmente bem com componentes de classe e funcionais. E por ser uma função pura, pode ser combinado com outros HOCs, ou até com si próprio.
```

Você deve ter notado similaridades entre HOCs e um padrão chamado **componentes container**. Componentes container são parte de uma estratégia de separação de responsabilidade entre preocupações de alto nível e baixo nível. Containers gerenciam coisas como subscrições e state, e passa props para componentes que lidam com coisas como renderização da UI. HOCs utilizam containers como parte de sua implementação. Você pode pensar em HOCs como definições de componentes container com parâmetros.

> ## Convenção: Passe Props Não Relacionadas Para o Componente Envolvido {#convention-pass-unrelated-props-through-to-the-wrapped-component}

Reviewer: ```suggestion
## Convenção: Passe Props Não Relacionadas Para o Componente Encapsulado {#convention-pass-unrelated-props-through-to-the-wrapped-component}
```

HOCs adicionam características a um componente. Eles não devem alterar drasticamente o seu contrato. É esperado que um componente retornado de um HOC tenha uma interface similar ao do componente envolvido.

HOCs devem passar props que não são relacionadas às suas preocupações específicas. A maioria dos HOCs possuem um método render que se parece com algo assim:

```js
render() {
  // Filtra props extras que são específicas a esse HOC e não devem ser

>   // passadas para o componente envolvido.

Reviewer: ```suggestion
  // passadas para o componente encapsulado.
```

  const { extraProp, ...passThroughProps } = this.props;

>   // Injeta props no componente envolvido.

Reviewer: ```suggestion
  // Injeta props no componente encapsulado.
```

  // Estes geralmente são valores do state ou métodos da instância.
  const injectedProp = someStateOrInstanceMethod;

>   // Passa as props para o componente envolvido.

Reviewer: ```suggestion
  // Passa as props para o componente encapsulado.
```

}
```

Essa convenção ajuda a garantir que HOCs são tão flexíveis e reutilizáves quanto possível.

## Convenção: Maximizando Composabilidade {#convention-maximizing-composability}

> Nem todos os HOCs são iguais. Às vezes eles aceitam apenas um único argumento, o componente envolvido:

Reviewer: ```suggestion
Nem todos os HOCs são iguais. Às vezes eles aceitam apenas um único argumento, o componente encapsulado:
```

```js
const NavbarWithRouter = withRouter(Navbar);
```

Normalmente, HOCs aceitam argumentos adicionais. Neste exemplo do Relay, um objeto de configuração é usado para especificar uma dependência de dados de um componente:

```js
const CommentWithRelay = Relay.createContainer(Comment, config);
```

A assinatura mais comum para HOCs se parece com isso:

```js
// `connect` do React Redux
const ConnectedComment = connect(commentSelector, commentActions)(CommentList);
```

*O Que?!* Se você quebrar isso em partes, é mais fácil de ver o que está acontecendo.

```js
// connect é uma função que retorna outra função
const enhance = connect(commentListSelector, commentListActions);

> // A função retornada é um HOC, o qual retorna um componente que está conectado

Reviewer: ```suggestion
// A função retornada é um HOC, que retorna um componente que está conectado
```

// à store do Redux
const ConnectedComment = enhance(CommentList);
```

Em outras palavras, `connect` é uma função de ordem superior que retorna um componente de ordem superior!

> Essa forma pode parecer confusa ou desnecessária, mas ela possui uma propriedade útil. HOCs com um único argumento, como o retornado pela função `connect`, possuem a assinatura `Componente => Componente`. Funções cujo tipo de saída é igual ao tipo da entrada são muito fáceis de compor juntas.

Reviewer: ```suggestion
Essa forma pode parecer confusa ou desnecessária, mas ela possui uma propriedade útil. HOCs com um único argumento, como o retornado pela função `connect`, possuem a assinatura `Componente => Componente`. Funções cujo tipo de saída é igual ao tipo da entrada são muito fáceis de se compor.
```

```js

> // Ao invés de fazer isso...

Reviewer: ```suggestion
// Em vez de fazer isso...
```

const EnhancedComponent = withRouter(connect(commentSelector)(WrappedComponent))

> // ... Você pode usar uma utilidade de composição de funções

Reviewer: ```suggestion
// ... Você pode usar uma conveniência de composição de funções
```

// compose(f, g, h) é o mesmo que (...args) => f(g(h(...args)))
const enhance = compose(
  // Ambos são HOCs com um único argumento
  withRouter,
  connect(commentSelector)
)
const EnhancedComponent = enhance(WrappedComponent)
```

(Essa mesma propriedade também permite que `connect` e outros enhancer-style HOCs sejam usados como decoradores, uma proposta experimental do JavaScript.)

> A função utilitária `compose` é providenciada por várias bibliotecas de terceiros, incluindo lodash (como o [`lodash.flowRight`](https://lodash.com/docs/#flowRight)), [Redux](http://redux.js.org/docs/api/compose.html), e [Ramda](http://ramdajs.com/docs/#compose).

Reviewer: ```suggestion
A função de conveniência `compose` é disponibilizada por várias bibliotecas de terceiros, incluindo lodash (como o [`lodash.flowRight`](https://lodash.com/docs/#flowRight)), [Redux](http://redux.js.org/docs/api/compose.html), e [Ramda](http://ramdajs.com/docs/#compose).
```

## Convenção: Envolva o Nome de Exibição para Depurar Melhor {#convention-wrap-the-display-name-for-easy-debugging}

Os componentes container criados por HOCs aparecem no [React Developer Tools](https://github.com/facebook/react-devtools) como qualquer outro componente. Para facilitar a depuração, escolha um nome de exibição que comunique que o componente é o resultado de um HOC.

> A técnica mais comum é envolver o nome de exibição do componente envolvido. Então, se o seu componente de ordem superior se chama `withSubscription`, e o nome de exibição do componente envolvido é `CommentList`, use o nome de exibição `WithSubscription(CommentList)`:

Reviewer: ```suggestion
A técnica mais comum é encapsular o nome de exibição do componente encapsulado. Então, se o seu componente de ordem superior se chama `withSubscription`, e o nome de exibição do componente encapsulado é `CommentList`, use o nome de exibição `WithSubscription(CommentList)`:
```

}
```

> ## Advertências {#caveats}

Reviewer: ```suggestion
## Ressalvas {#caveats}
```

> Componentes de ordem superior vêm com algumas advertências que não são imediatamente óbvias se você for novo no React.

Reviewer: ```suggestion
Componentes de ordem superior vêm com algumas ressalvas que não são imediatamente óbvias se você for novo no React.
```

### Não Use HOCs Dentro do Método render {#dont-use-hocs-inside-the-render-method}

O algoritmo de comparação do React (chamado reconciliação) usa a identidade do componente para determinar se ele deve atualizar a sub-árvore existente ou jogá-la fora e montar uma nova. Se o componente que foi retornado do `render` é idêntico (`===`) ao componente do render anterior, o React atualizará a sub-árvore comparando-a recursivamente com uma nova. Se elas não forem iguais, a sub-árvore anterior será completamente desmontada.

Normalmente, você não precisa se preocupar com isso. Mas isso é importante em HOCs porque não se pode aplicar um HOC em um componente dentro do método render de um componente:

```js
render() {
  // Uma nova versão de EnhancedComponent é criada toda vez que render for chamado
  // EnhancedComponent1 !== EnhancedComponent2
  const EnhancedComponent = enhance(MyComponent);
  // Isso faz com que a sub-árvore seja completamente
  // desmontada/remontada todas as vezes!
  return <EnhancedComponent />;
}
```

O problema aqui não é só na performance - remontar um componente causa a perda de todo o seu state e todos os seus filhos.

> Ao invés disso, aplique HOCs fora da definição do componente para que o componente resultante seja criado apenas uma vez. Então, sua identidade será consistente pelas renderizações. De qualquer forma, isso geralmente é o que você quer.

Reviewer: ```suggestion
Em vez disso, aplique HOCs fora da definição do componente para que o componente resultante seja criado apenas uma vez. Então, sua identidade será consistente pelas renderizações. De qualquer forma, isso geralmente é o que você quer.
```

> Nesses casos raros em que é preciso aplicar um HOC dinamicamente, isso também pode ser feito dentro dos métodos de ciclo de vida do componente, ou no seu constructor.

Reviewer 4: ```suggestion
Nesses casos raros em que é preciso aplicar um HOC dinamicamente, isso também pode ser feito dentro dos métodos de ciclo de vida do componente, ou no seu construtor.
```

### Métodos estáticos devem ser copiados {#static-methods-must-be-copied-over}

As vezes é útil definir um método estático em um componente do React. Por exemplo, containers do Relay expõem um método estático `getFragment` para facilitar a composição de fragmentos do GraphQL.

> Mas, quando se aplica um HOC a um componente, o componente original é envolvido com um componente container. Isso significa que o novo componente não possui nenhum dos métodos estáticos do componente original.

Reviewer: ```suggestion
Mas, quando se aplica um HOC a um componente, o componente original é encapsulado com um componente container. Isso significa que o novo componente não possui nenhum dos métodos estáticos do componente original.
```

}
```

Outra solução possível é exportar o método estático do próprio componente, separadamente.

```js

> // Ao invés de...

Reviewer: ```suggestion
// Em vez de...
```

MyComponent.someFunction = someFunction;
export default MyComponent;

// ... exporte o método separadamente...
export { someFunction };

// ... e no módulo que está consumindo, importe ambos
import MyComponent, { someFunction } from './MyComponent.js';
```

> ### Refs Não São Passados Diretamente {#refs-arent-passed-through}

Reviewer 4: ```suggestion
### Refs Não São Passadas Diretamente {#refs-arent-passed-through}
```

> Enquanto a convenção para componentes de ordem superior é passar diretamente todas as props para o componente envolvido, isso não funciona para refs. Isso acontece porque `ref` não é exatamente uma prop - como `key`, ele é lidado especificamente pelo React. Se um ref for adicionado a um elemento cujo componente é o resultado de um HOC, o ref referenciará à instância mais externa do componente container, não o componente envolvido.

Reviewer 4: ```suggestion
Enquanto a convenção para componentes de ordem superior é passar diretamente todas as props para o componente envolvido, isso não funciona para refs. Isso acontece porque `ref` não é exatamente uma prop - como `key`, ele é lidado especificamente pelo React. Se uma ref for adicionada a um elemento cujo componente é o resultado de um HOC, a ref referenciará à instância mais externa do componente container, não o componente envolvido.
```

> Enquanto a convenção para componentes de ordem superior é passar diretamente todas as props para o componente envolvido, isso não funciona para refs. Isso acontece porque `ref` não é exatamente uma prop - como `key`, ele é lidado especificamente pelo React. Se um ref for adicionado a um elemento cujo componente é o resultado de um HOC, o ref referenciará à instância mais externa do componente container, não o componente envolvido.

Reviewer: ```suggestion
Enquanto a convenção para componentes de ordem superior é passar diretamente todas as props para o componente encapsulado, isso não funciona para refs. Isso acontece porque `ref` não é exatamente uma prop - tal como `key`, ele é tratado de maneira especial pelo React. Se um ref for adicionado a um elemento cujo componente é o resultado de um HOC, o ref referenciará à instância mais externa do componente container, não ao componente encapsulado.
```
